pulp_puppet_plugins/pulp_puppet/plugins/db/models.py

```diff
@@ -0,0 +1,196 @@
+from mongoengine import ListField, StringField
+from pulp.common.compat import json
+from pulp.server.db.model import FileContentUnit
+
+from pulp_puppet.common import constants
+from pulp_puppet.plugins.importers import metadata as metadata_parser
+
+
+class RepositoryMetadata(object):
+    """
+    An object that stores and produces Puppet Repository metadata
+
+    :ivar modules: the list of modules in the repository
+    :type modules: list
+    """
+
+    def __init__(self):
+        self.modules = []
+
+    def update_from_json(self, metadata_json):
+        """
+        Updates this metadata instance with modules found in the given JSON
+        document. This can be called multiple times to merge multiple
+        repository metadata JSON documents into this instance.
+
+        :return: object representing the repository and all of its modules
+        :rtype:  RepositoryMetadata
+        """
+
+        parsed = json.loads(metadata_json)
+
+        # The contents of the metadata document is a list of dictionaries,
+        # each representing a single module.
+        for module_dict in parsed:
+            module = Module.from_metadata(module_dict)
+            self.modules.append(module)
+
+    def to_json(self):
+        """
+        Return the repository metadata as a JSON representation.
+
+        :return: The repository metadata as json.
+        :rtype: str
+        """
+        repo_metadata_dict = []
+        for module in self.modules:
+            module_metadata = {'name': module.name, 'author': module.author,
+                               'version': module.version, 'tag_list': module.tag_list}
+            repo_metadata_dict.append(module_metadata)
+
+        # Serialize metadata of all modules in the repo into a single JSON document
+        return json.dumps(repo_metadata_dict)
+
+
+class Module(FileContentUnit):
+    """
+    The mongoengine representation of a Puppet Module.
+    """
```

Comment:
I purposely didn't document these because they are all documented by the Puppet metadata documentation. I did add a note in the class docstring that all fields correspond with Puppet metadata attributes and users should refer to those docs for the field documentation.